### PR TITLE
SQL type conversion details

### LINKED ISSUE #51: sql: type conversion
Should be part of the SQL Reference section.

### LINKED ISSUE #282: Document which types can be cast into others
Feedback from user in Gitter: Functions and operators or the Data Types pages should have a listing of what types can be cast into others.

### PR DESCRIPTION
**Pull request by sploiselle**
Closes #51
Closes #282

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/977)

### CODE REVIEW

**Comment by knz** [comment]
I do not like and neither agree with the notion of "unsupported casts".

The idea of a cast function in the first place is a "bonus" and there is no moral mandate to make casts from anything to anything exist.
The expected, normal baseline state of affairs for any computer language, including SQL, is *when no cast function exists at all*. Your tables suggest that the expected state is one where all casts are possible, and that us not supporting some of them is an embarrassing, temporary state of affairs which we intend to escape someday. That's not a good message to send.

May I suggest instead removing the "unsupported" rows entirely from these table, as well as the "supported" column, and the mention about SERIAL below each of them. 

Then on each line in the details section in each table give an example of the possible values before and after each conversion. This is especially useful to show what happens when converting to/from numeric and time types.

---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[bool.md, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/977#-Ka21kr3qkw4KZ4kt8FU:-Ka21kr4shG_D320vwhE:b-kj0ink) ([raw file](https://github.com/cockroachdb/docs/blob/5ef21b525a080e9edd6500c5258467ebc1582d71/bool.md#L59)):*
> ```Markdown
> ## Supported Casting & Conversion
> 
> `BOOL` values support the following [data-type casting](data-types.html#type-casting--conversion):
> ```

Can you clarify here (and below) the direction of the cast: "The following table summarizes the types that `BOOL` values can be [casted](...) to:"

---

*[data-types.md, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/977#-Ka23Y5BLyPD3HpiQmTO:-Ka23Y5Cf5HWZUgUfwN0:bt185vc) ([raw file](https://github.com/cockroachdb/docs/blob/5ef21b525a080e9edd6500c5258467ebc1582d71/data-types.md#L24)):*
> ```Markdown
> [`BYTES`](bytes.html) | A string of binary characters. | `b'\141\061\142\062\143\063'`
> 
> ## Type Casting & Conversion
> ```

"Data type conversions & casts"

---

*[data-types.md, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/977#-Ka234RoER9zcsZmWj10:-Ka234RoER9zcsZmWj11:bcfd4ew) ([raw file](https://github.com/cockroachdb/docs/blob/5ef21b525a080e9edd6500c5258467ebc1582d71/data-types.md#L26)):*
> ```Markdown
> ## Type Casting & Conversion
> 
> CockroachDB supports casting one data type into another using the following methods:
> ```

"CockroachDB supports explicit type conversions using the following methods:

1. `<type> 'string literal'`, to convert from the literal representation of a value to a value of that type. For example:
     `DATE '2008-12-21'`, `INT '123'` or `BOOL 'true'`

2. `<Value>::<data type>`, or its equivalent longer form `CAST(<value> AS <data type>)`, which converts an arbitrary expression of one built-in type to another (this is also known as "casting"). For example:
     `NOW()::DECIMAL`, `VARIANCE(a+2)::INT`

3. Other built-in conversion functions when the type is not a SQL type, for example `FROM_IP()` / `TO_IP()` (from/to IP addresses), `FROM_UUID()`/`TO_UUID()` (from/to UUIDs).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/977#-:-Ka25z86PW3SJvla_zPi:bm1clmm)*

**Comment by knz** [comment]
Also add a link in the "Type Casting & Conversion" to the page that documents ANNOTATE_TYPE and the `:::` operator. In many cases `:::` should be preferred to `::` as it is faster to evaluate while executing the query.

**Comment by sploiselle** [comment]
@knz Do you think the work to describe `ANNOTATE_TYPE` (so we can link to it) is covered in issue #346 ? If so, I'm glad to add that as a follow-on bit of work, but would like to close out this PR before taking that on because it's an indeterminate amount of work at this point.

---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[bool.md, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/977#-Ka21kr3qkw4KZ4kt8FU:-KaDDzl6MEqUeKDaXQuS:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/5ef21b525a080e9edd6500c5258467ebc1582d71/bool.md#L59)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you clarify here (and below) the direction of the cast: "The following table summarizes the types that `BOOL` values can be [casted](...) to:"
</blockquote></details>

Done.

---

*[data-types.md, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/977#-Ka23Y5BLyPD3HpiQmTO:-KaDEAWdqXrsEVOEjnCU:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/5ef21b525a080e9edd6500c5258467ebc1582d71/data-types.md#L24)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"Data type conversions & casts"
</blockquote></details>

Done.

---

*[data-types.md, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/977#-Ka234RoER9zcsZmWj10:-KaE9FjW0snPoZtt-fwq:beolbqx) ([raw file](https://github.com/cockroachdb/docs/blob/5ef21b525a080e9edd6500c5258467ebc1582d71/data-types.md#L26)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"CockroachDB supports explicit type conversions using the following methods:

1. `<type> 'string literal'`, to convert from the literal representation of a value to a value of that type. For example:
     `DATE '2008-12-21'`, `INT '123'` or `BOOL 'true'`

2. `<Value>::<data type>`, or its equivalent longer form `CAST(<value> AS <data type>)`, which converts an arbitrary expression of one built-in type to another (this is also known as "casting"). For example:
     `NOW()::DECIMAL`, `VARIANCE(a+2)::INT`

3. Other built-in conversion functions when the type is not a SQL type, for example `FROM_IP()` / `TO_IP()` (from/to IP addresses), `FROM_UUID()`/`TO_UUID()` (from/to UUIDs).


</blockquote></details>

Ah––this is perfect! Thank you for doing the writing for me, Raphael!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/977#-:-KaEDBbaUoTcmGDE3P8S:b-fva3to)*

**Comment by knz** [comment]
annotate_type is not a function so I don't think it belongs with #346.

---

Reviewed 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[bytes.md, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEY2VcDvrD2ghxyXXE:-KaEY2Vdd91wardpVcFE:b-r5gcjy) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/bytes.md#L59)):*
> ```Markdown
> Type | Details
> -----|--------
> `STRING` | ––
> ```

Details: the conversion is only possible if the byte array contains valid UTF-8 character encodings.

---

*[data-types.md, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/977#-Ka234RoER9zcsZmWj10:-KaEYArxOIWytSK7pT2k:bb6mt9b) ([raw file](https://github.com/cockroachdb/docs/blob/5ef21b525a080e9edd6500c5258467ebc1582d71/data-types.md#L26)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary><blockquote>

Ah––this is perfect! Thank you for doing the writing for me, Raphael!
</blockquote></details>

:+1:

---

*[date.md, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEYEgtLS0hEajF3yyK:-KaEYEgtLS0hEajF3yyL:bagpxca) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/date.md#L60)):*
> ```Markdown
> `DECIMAL` | Converts to number of days since the Unix epoch (Jan. 1, 1970)
> `FLOAT` | Converts to number of days since the Unix epoch (Jan. 1, 1970)
> `TIMESTAMP` | ––
> ```

Details: Sets the time to 00:00 in the resulting timestamp.

---

*[decimal.md, line 78 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEYNzcAY1tew6i9YQl:-KaEYNzcAY1tew6i9YQm:b-477h0l) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/decimal.md#L78)):*
> ```Markdown
> -----|--------
> `INT` | Truncates decimal precision
> `FLOAT` | ––
> ```

Details: loses precision, and may round up to +/- infinity if the value is too large in magnitude, or to +/-0 if the value is too small in magnitude.

---

*[float.md, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEY_sCsC9a4CUZuqSE:-KaEY_sCsC9a4CUZuqSF:b5p2p2c) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/float.md#L68)):*
> ```Markdown
> Type | Details
> -----|--------
> `INT` | Truncates decimal precision
> ```

... also the behavior is undefined if the value is larger than +2^63-1 or smaller than -2^63

---

*[int.md, line 77 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEYvtzlHSIZcwtLfwQ:-KaEYvu-quHzMj3LZMUL:b-jjz02k) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/int.md#L77)):*
> ```Markdown
> -----|--------
> `DECIMAL` | ––
> `FLOAT` | ––
> ```

Detail: will be inaccurate for values larger than 2^53 in magnitude.

---

*[serial.md, line 111 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEZHEPsHcXPGAJ3BTE:-KaEZHEPsHcXPGAJ3BTF:bq9qjlu) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/serial.md#L111)):*
> ```Markdown
> ## Supported Casting & Conversion
> 
> Because the `SERIAL` data type represents values automatically generated CockroachDB to uniquely identify rows, you cannot meaningfully cast `SERIAL` into other data types.
> ```

That's not quite correct I think, IIRC "serial" is just INT in disguise.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/977#-:-KaEZVFEbX5ZzEcP71MN:btgvp54)*

**Comment by sploiselle** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[bytes.md, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEY2VcDvrD2ghxyXXE:-KaJ4Psv-o8khpPp9udS:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/bytes.md#L59)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Details: the conversion is only possible if the byte array contains valid UTF-8 character encodings.
</blockquote></details>

Done.

---

*[date.md, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEYEgtLS0hEajF3yyK:-KaJ12Bp1xbtZh4B2jmA:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/date.md#L60)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Details: Sets the time to 00:00 in the resulting timestamp.
</blockquote></details>

Done.

---

*[decimal.md, line 78 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEYNzcAY1tew6i9YQl:-KaJ1AFGtSdWrLhfRhj4:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/decimal.md#L78)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Details: loses precision, and may round up to +/- infinity if the value is too large in magnitude, or to +/-0 if the value is too small in magnitude.
</blockquote></details>

Done.

---

*[float.md, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEY_sCsC9a4CUZuqSE:-KaJ1QTUdQobSrL3Yqyg:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/float.md#L68)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

... also the behavior is undefined if the value is larger than +2^63-1 or smaller than -2^63
</blockquote></details>

Done.

---

*[int.md, line 77 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEYvtzlHSIZcwtLfwQ:-KaJ1a1d7YpJ3csW3SZ5:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/int.md#L77)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Detail: will be inaccurate for values larger than 2^53 in magnitude.
</blockquote></details>

Done.

---

*[serial.md, line 111 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/977#-KaEZHEPsHcXPGAJ3BTE:-KaJ4OOt5atD400XbxYt:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2d3c07f95cc8653a06bc1d0423ff0ddd2333c87a/serial.md#L111)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

That's not quite correct I think, IIRC "serial" is just INT in disguise.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/977)*

**Comment by knz** [comment]
:lgtm:

---

Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/977#-:-KaJ7mfE5BNEYw05KbbQ:bnfp4nl)*

File: timestamp.md [hunk 1]
```diff
@@ -63,6 +63,21 @@ A `TIMESTAMP` column supports values up to 12 bytes in width, but the total stor
 # Note that the first timestamp is UTC-05:00, which is the equivalent of EST.
 ~~~
 
+## Supported Casting & Conversion
+
+`TIMESTAMP` values can be cast to any of the following data types:
+
+Type | Details
+-----|--------
+`INT | Converts to number of seconds since the Unix epoch (Jan. 1, 1970)
```
Comment (jseldess) [inline timestamp.md]: Missing closing back-ticks. But I want to get this in the release notes, so I'll merge and correct in a follow up.

**Pull request merged**